File: HavenApp/HavenApp/Services/EngagementTracker.swift
import Foundation

/// Pure engagement bookkeeping — likes, zaps, reposts, and per-note stats.
/// No Combine, no @Published. FeedService delegates state mutations here
/// and assigns results to its published properties.
enum EngagementTracker {

    // MARK: - Interaction State (Persistence Model)

    /// Codable snapshot of the user's engagement state (which notes they've
    /// liked/zapped), persisted to disk per account.
    struct InteractionState: Codable {
        let likedEventIds: Set<String>
        let zappedEventIds: [String: Int]

        init(likedEventIds: Set<String>, zappedEventIds: [String: Int]) {
            self.likedEventIds = likedEventIds
            self.zappedEventIds = zappedEventIds
        }

        init(from decoder: Decoder) throws {
            let container = try decoder.container(keyedBy: CodingKeys.self)
            likedEventIds = try container.decode(Set<String>.self, forKey: .likedEventIds)
            // Migrate from old Set<String> format if needed
            if let dict = try? container.decode([String: Int].self, forKey: .zappedEventIds) {
                zappedEventIds = dict
            } else if let set = try? container.decode(Set<String>.self, forKey: .zappedEventIds) {
                zappedEventIds = Dictionary(uniqueKeysWithValues: set.map { ($0, 0) })
            } else {
                zappedEventIds = [:]
            }
        }
    }

    // MARK: - File Paths

    /// Returns the per-account interaction state file URL.
    static func interactionStateURL(forKey key: String) -> URL {
        let havenDir = havenSupportDir()
        let safeKey = key.isEmpty ? "owner" : key.replacingOccurrences(of: "/", with: "_")
        return havenDir.appendingPathComponent("interaction_state_\(safeKey).json")
    }

    /// Returns the legacy (pre-per-account) interaction state file URL.
    static func legacyInteractionStateURL() -> URL {
        return havenSupportDir().appendingPathComponent("interaction_state.json")
    }

    private static func havenSupportDir() -> URL {
        guard let appSupport = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first else {
            return URL(fileURLWithPath: NSTemporaryDirectory()).appendingPathComponent("Haven")
        }
        let havenDir = appSupport.appendingPathComponent("Haven", isDirectory: true)
        try? FileManager.default.createDirectory(at: havenDir, withIntermediateDirectories: true)
        return havenDir
    }

    // MARK: - Load / Save

    /// Loads interaction state from disk for the given account key.
    /// Falls back to the legacy global file when no per-account file exists.
    static func loadInteractionState(forKey key: String) -> (likedEventIds: Set<String>, zappedEventIds: [String: Int]) {
        let url = interactionStateURL(forKey: key)
        if let data = try? Data(contentsOf: url),
           let state = try? JSONDecoder().decode(InteractionState.self, from: data) {
            return (state.likedEventIds, state.zappedEventIds)
        }

        // Migration: read the old shared file
        let legacy = legacyInteractionStateURL()
        if let data = try? Data(contentsOf: legacy),
           let state = try? JSONDecoder().decode(InteractionState.self, from: data) {
            return (state.likedEventIds, state.zappedEventIds)
        }

        return ([], [:])
    }

    /// Persists interaction state to disk. Runs the encode + write on a
    /// background queue to avoid blocking the main thread.
    static func saveInteractionState(
        likedEventIds: Set<String>,
        zappedEventIds: [String: Int],
        forKey key: String
    ) {
        let state = InteractionState(likedEventIds: likedEventIds, zappedEventIds: zappedEventIds)
        let url = interactionStateURL(forKey: key)
        DispatchQueue.global(qos: .utility).async {
            if let data = try? JSONEncoder().encode(state) {
                try? data.write(to: url)
            }
        }
    }

    // MARK: - Self-Like Detection

    /// Identifies reactions authored by the owner in a batch of reaction events.
    /// Returns the note IDs that the owner has liked.
    static func detectSelfLikes(
        reactions: [(targetId: String, pubkey: String)],
        ownerHex: String
    ) -> [String] {
        guard !ownerHex.isEmpty else { return [] }
        return reactions.compactMap { $0.pubkey == ownerHex ? $0.targetId : nil }
    }

    // MARK: - Engagement Count Merging

    /// Merges a batch of reaction events and repost targets into the running
    /// per-note stats dictionary. Returns the updated dictionary.
    static func mergeEngagementCounts(
        reactions: [(targetId: String, pubkey: String)],
        repostTargets: [String],
        currentStats: [String: NoteStats]
    ) -> [String: NoteStats] {
        var updated = currentStats
        for (targetId, _) in reactions {
            var s = updated[targetId] ?? NoteStats()
            s.reactions += 1
            updated[targetId] = s
        }
        for targetId in repostTargets {
            var s = updated[targetId] ?? NoteStats()
            s.reposts += 1
            updated[targetId] = s
        }
        return updated
    }
}

File: NostrVault/app/src/main/java/com/nostrvault/data/local/EngagementTracker.kt
package com.nostrvault.data.local

import android.content.Context
import com.nostrvault.data.model.NoteStats
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import java.io.File

/**
 * Port of EngagementTracker.swift -- per-account like/zap state persistence
 * and engagement count merging. Pure logic, no UI dependencies.
 */
object EngagementTracker {

    private val json = Json { ignoreUnknownKeys = true }
    private lateinit var dataDir: File

    fun init(context: Context) {
        dataDir = File(context.filesDir, "nostrvault_data")
        dataDir.mkdirs()
    }

    // ---- Persistence model ----

    @Serializable
    data class InteractionState(
        val likedEventIds: Set<String>,
        val zappedEventIds: Map<String, Int>,
    )

    // ---- File paths ----

    private fun interactionStateFile(key: String): File {
        val safeKey = key.ifEmpty { "owner" }.replace("/", "_")
        return File(dataDir, "interaction_state_$safeKey.json")
    }

    private val legacyFile get() = File(dataDir, "interaction_state.json")

    // ---- Load / Save ----

    /**
     * Load interaction state from disk for the given account key.
     * Falls back to the legacy global file when no per-account file exists.
     */
    fun loadInteractionState(forKey: String): Pair<Set<String>, Map<String, Int>> {
        val file = interactionStateFile(forKey)

        // Try per-account file
        val state = loadFromFile(file) ?: loadFromFile(legacyFile)
        return if (state != null) {
            state.likedEventIds to state.zappedEventIds
        } else {
            emptySet<String>() to emptyMap()
        }
    }

    /**
     * Persist interaction state to disk on a background thread.
     */
    suspend fun saveInteractionState(
        likedEventIds: Set<String>,
        zappedEventIds: Map<String, Int>,
        forKey: String,
    ) = withContext(Dispatchers.IO) {
        val state = InteractionState(likedEventIds, zappedEventIds)
        val file = interactionStateFile(forKey)
        try {
            dataDir.mkdirs()
            file.writeText(json.encodeToString(state))
        } catch (_: Exception) { }
    }

    // ---- Self-like detection ----

    /**
     * Identify reactions authored by the owner in a batch of reaction events.
     * Returns the note IDs that the owner has liked.
     */
    fun detectSelfLikes(
        reactions: List<Pair<String, String>>, // (targetId, pubkey)
        ownerHex: String,
    ): List<String> {
        if (ownerHex.isEmpty()) return emptyList()
        return reactions.filter { it.second == ownerHex }.map { it.first }
    }

    // ---- Engagement count merging ----

    /**
     * Merge a batch of reaction events and repost targets into the running
     * per-note stats dictionary. Returns the updated dictionary.
     */
    fun mergeEngagementCounts(
        reactions: List<Pair<String, String>>, // (targetId, pubkey)
        repostTargets: List<String>,
        currentStats: Map<String, NoteStats>,
    ): Map<String, NoteStats> {
        val updated = currentStats.toMutableMap()

        for ((targetId, _) in reactions) {
            val stats = updated.getOrPut(targetId) { NoteStats() }
            stats.reactionCount++
            updated[targetId] = stats
        }
        for (targetId in repostTargets) {
            val stats = updated.getOrPut(targetId) { NoteStats() }
            stats.repostCount++
            updated[targetId] = stats
        }

        return updated
    }

    // ---- Private ----

    private fun loadFromFile(file: File): InteractionState? = try {
        if (file.exists()) json.decodeFromString(file.readText()) else null
    } catch (_: Exception) {
        null
    }
}
